File: backend/go/dao/post-dao.go
package main

import (
	"dao/models"
	"errors"
	"gorm.io/gorm"
)

var IllegalArgumentError = errors.New("illegal argument error")

type PostDAO interface {
	FindAll() ([]*models.Post, error)
	Create(post *models.Post) error
	Update(post *models.Post) error
	Delete(id int64) error
}

type postDAO struct {
	db *gorm.DB
}

func (dao *postDAO) FindAll() ([]*models.Post, error) {
	var posts []*models.Post
	result := dao.db.Find(&posts)

	if result.Error != nil {
		return nil, result.Error
	}
	return posts, nil
}

func (dao *postDAO) Create(post *models.Post) error {
	if post.ID != 0 {
		return IllegalArgumentError
	}

	result := dao.db.Create(post)
	return result.Error
}

func (dao *postDAO) Update(post *models.Post) error {
	fetchedPost := &models.Post{}
	result := dao.db.Find(fetchedPost, post.ID)
	if result.RowsAffected == 0 {
		return IllegalArgumentError
	}

	result = dao.db.Updates(post)
	return result.Error
}

func (dao *postDAO) Delete(id int64) error {
	return errors.New("not implemented")
}

File: backend/python/dao/src/PostDAO.py
from sqlalchemy.orm import sessionmaker
import Post
from Exception import IllegalArgumentException, UnknownElementException

def takeId(elem):
  return elem.id

class PostDAO(object):
  def __init__(self, engine, schema = None):
    self.engine = engine
    Session = sessionmaker(bind = engine)
    self.session = Session()
  
  def findAll(self):
    return sorted(self.session.query(Post.Post).all(), key=takeId)

  def create(self, post):
    if (post.id is not None):
      raise IllegalArgumentException()
    self.session.add(post)
    self.session.flush()

  def update(self, post):
    fetchedPost = self.session.query(Post.Post).get(post.id)
    if (fetchedPost is None):
      raise UnknownElementException()
    self.session.merge(post)
    self.session.commit()
    self.session.flush()
    fetchedPost3 = self.session.query(Post.Post).get(post.id)

  def delete(self, postId):
    fetchedPost = self.session.query(Post.Post).get(postId)
    if (fetchedPost is None):
      raise UnknownElementException()
    self.session.delete(fetchedPost)
    self.session.commit()
    self.session.flush()